packages/gdl-frontend/components/OnlineStatusRedirectProvider.js

```diff
@@ -0,0 +1,95 @@
+// @flow
+import * as React from 'react';
+import { withRouter } from 'next/router';
+import offlineLibrary from '../lib/offlineLibrary';
+import OnlineStatusContext from './OnlineStatusContext';
+
+/**
+ * Watches the user's online status and provides it as context.
+ * Redirects to the offline library if deemed necessary
+ */
+class OnlineStatusRedirectProvider extends React.Component<
+  {
+    children: React.Node,
+    router: {
+      replace: string => void,
+      route: string,
+      query: { id?: string, lang?: string }
+    }
+  },
+  { online: boolean }
+> {
+  state = {
+    online: getStatus()
+  };
+
+  handleOnline = () => this.setState({ online: true });
+
+  /**
+   * When we go offline, we redirect to the offline library page (unless we are at a book that is already offlined)
+   */
+  handleOffline = async () => {
+    // Only redirect to the offline page if we have proper offline support (We don't want to redirect in IE for instance)
+    if (!offlineLibrary) {
+      return;
+    }
+
+    this.setState({ online: false });
+    const {
+      router,
+      router: { route, query }
+    } = this.props;
+
+    if (
+      (route === '/books/_read' || route === '/books/_book') &&
```

Comment:
I assume the `translate` route should not be considered here? When a user is in `/translate` and goes offline, they should be prompted to `/offline`, right?